TDataFrame.hxx

```diff
@@ -22,828 +22,1042 @@
+         auto title    = branch->GetTitle();
+         auto typeCode = title[strlen(title) - 1];
+         if (typeCode == 'B') {
+            return SimpleAction<Char_t, ART_t, at, TT_t>::BuildAndBook(this, theBranchName, r, nSlots);
+         }
+         // else if (typeCode == 'b') { return SimpleAction<UChar_t, ART_t, at, TT_t>::BuildAndBook(this, theBranchName, r, nSlots); }
+         // else if (typeCode == 'S') { return SimpleAction<Short_t, ART_t, at, TT_t>::BuildAndBook(this, theBranchName, r, nSlots); }
+         // else if (typeCode == 's') { return SimpleAction<UShort_t, ART_t, at, TT_t>::BuildAndBook(this, theBranchName, r, nSlots); }
+         else if (typeCode == 'I') {
+            return SimpleAction<int, ART_t, at, TT_t>::BuildAndBook(this, theBranchName, r, nSlots);
+         }
+         // else if (typeCode == 'i') { return SimpleAction<unsigned int , ART_t, at, TT_t>::BuildAndBook(this, theBranchName, r, nSlots); }
+         // else if (typeCode == 'F') { return SimpleAction<Float_t, ART_t, at, TT_t>::BuildAndBook(this, theBranchName, r, nSlots); }
+         else if (typeCode == 'D') {
+            return SimpleAction<Double_t, ART_t, at, TT_t>::BuildAndBook(this, theBranchName, r, nSlots);
+         }
+         // else if (typeCode == 'L') { return SimpleAction<Long64_t, ART_t, at, TT_t>::BuildAndBook(this, theBranchName, r, nSlots); }
+         // else if (typeCode == 'l') { return SimpleAction<ULong64_t, ART_t, at, TT_t>::BuildAndBook(this, theBranchName, r, nSlots); }
+         else if (typeCode == 'O') {
+            return SimpleAction<bool, ART_t, at, TT_t>::BuildAndBook(this, theBranchName, r, nSlots);
+         }
       } else {
          std::string typeName = branchEl->GetTypeName();
-         if (typeName == "vector<double>") { return SimpleAction<std::vector<double>, ART, AT, TT>::BuildAndBook(this, theBranchName, r, nSlots); }
-         else if (typeName == "vector<float>") { return SimpleAction<std::vector<float>, ART, AT, TT>::BuildAndBook(this, theBranchName, r, nSlots); }
-         else if (typeName == "vector<int>") { return SimpleAction<std::vector<int>, ART, AT, TT>::BuildAndBook(this, theBranchName, r, nSlots); }
+         if (typeName == "vector<double>") {
+            return SimpleAction<std::vector<Double_t>, ART_t, at, TT_t>::BuildAndBook(this, theBranchName, r, nSlots);
+         } else if (typeName == "vector<float>") {
+            return SimpleAction<std::vector<Float_t>, ART_t, at, TT_t>::BuildAndBook(this, theBranchName, r, nSlots);
+         }
       }
-      return SimpleAction<BranchType, ART, AT, TT>::BuildAndBook(this, theBranchName, r, nSlots);
+      return SimpleAction<BranchType, ART_t, at, TT_t>::BuildAndBook(this, theBranchName, r, nSlots);
    }
-
-   template<typename T>
-   void Book(std::unique_ptr<T> ptr) {
-      fDerivedPtr->Book(std::move(ptr));
+   template <typename T>
+   void Book(std::shared_ptr<T> ptr)
+   {
+      fProxiedPtr->Book(ptr);
```

Comment:
it would be better to call book directly on the `TDataFrameImpl` object and remove all dummy `Book` methods from `TDataFrame{Action, Branch, Filter}`.